clients/src/main/java/org/apache/kafka/common/internals/FetchBuilder.java

```diff
@@ -0,0 +1,170 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.common.internals;
+
+import org.apache.kafka.common.TopicPartition;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * This builder is a useful building block for doing fetch requests where topic partitions have to be rotated via
+ * round-robin to ensure fairness and some level of determinism given the existence of a limit on the fetch response
+ * size. Because the serialization of fetch requests is more efficient if all partitions for the same topic are grouped
+ * together, we do such grouping in the method `set`.
+ *
+ * As partitions are moved to the end, the same topic may be repeated more than once.
+ */
+public class FetchBuilder<S> {
+
+    private final LinkedHashMap<TopicPartition, S> map = new LinkedHashMap<>();
```

Comment:
Initially, when `set(partitionState)` is called, the partitions are ordered by topic. Over time, the order becomes more confusing since the partitions are fetched and rotated in different orders. As long as partitions stay assigned to the same brokers, the relative order should be preserved for each broker by topic. But as partitions move, it seems like we'll lose that relative ordering. I'm not sure how big of a concern this is since the ordering is just an optimization, but maybe we need to occasionally reset the state to avoid it becoming too disorganized. Either that, or we should do the round-robin per-broker.